### PR TITLE
Adding interactive measurement abilities to OpenSCAD

### LINKED ISSUE #3638: Provide a simple measurement tool [$805]
$500 to $1000 bounty depending on final capabilities of the tool.  Let me know how to set this up so you can verify the money is available and/or in some sort of escrow account.

I love OpenSCAD, but I do enough stuff where I need to I need to measure dimensions to make sure everything is OK that I can't really use OpenSCAD for what I want to do.

I am looking for a tool that does the following:
- Measure distances between two point
- Measure angles between three points (so essentially you create a triangle without the hypotenuse side and the angle between the two lines created by the three points is measured)
- Snap to edges, corners, and midpoints

I have not dug through OpenSCAD's source code to see what is possible, but if this is not infeasible due to the underlying structure of OpenSCAD, I would love to have this tool.

Let me know what you think, and/or if I'm publishing this feature request here correctly.  It has been almost ten years since I was a full time software engineer, and GitHub was in its infancy back then.  And, also, this is probably a dupe of this feature request, but I didn't want to bump an almost five year old feature request.

https://github.com/openscad/openscad/issues/1768

<bountysource-plugin>

---
There is a **[$805 open bounty](https://app.bountysource.com/issues/96259516-provide-a-simple-measurement-tool?utm_campaign=plugin&utm_content=tracker%2F52063&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://app.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F52063&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gsohler**
This PR enables PR to measure distances and angle measurement features to OpenSCAD, which is a long awaited feature.
For distance you need to enter 2 points, For angle you enter 3 points. 
Selecting a Point in the model works like raytracing. for each eye ray, the closest point is selected.
Key function for this ability is gluUnproject which basically does the reverse of displaying Points

See https://www.youtube.com/watch?v=xunM6tBu-E0 for demonstration.

Cheers

Fixes #3638.

### CODE REVIEW

**Comment by pca006132** [comment]
btw is the manifold submodule update intended? or should we wait for #4728?

**Comment by pythonscad** [comment]
I was working on your suggestions ...

https://www.youtube.com/watch?v=LA9lsDdwhTA

**Comment by gsohler** [comment]
This does not interact with  VBO.

Only contact point with GL is reading back the projection and model
matrices.

File: src/glview/GLView.h [hunk 1]
```diff
@@ -69,6 +71,10 @@ class GLView
   bool showedges;
   bool showcrosshairs;
   bool showscale;
+  GLdouble modelview[16];
+  GLdouble projection[16];
```
Comment (t-paul) [inline src/glview/GLView.h:75]: This causes the build fail for WASM / HEADLESS. The simple, slightly ugly way for solving the issue is probably another entry in the NULLGL section of system-gl.h.

File: src/glview/cgal/CGALRenderer.cc [hunk 2]
```diff
@@ -333,3 +333,81 @@ BoundingBox CGALRenderer::getBoundingBox() const
...
+}
+
+std::vector<SelectedObject> CGALRenderer::findModelObject(Vector3d near_pt, Vector3d far_pt,int mouse_x, int mouse_y, double tolerance) {
+  std::vector<SelectedObject> results;
+  double dist_near;
+  double dist_nearest=NAN;
+  Vector3d pt1_nearest;
+  Vector3d pt2_nearest;
+  for (const auto& p : this->getPolyhedrons()) {
+  }
```
Comment (t-paul) [inline src/glview/cgal/CGALRenderer.cc:365]: ?

File: src/glview/preview/OpenCSGRenderer.cc [hunk 3]
```diff
@@ -508,3 +508,13 @@ BoundingBox OpenCSGRenderer::getBoundingBox() const
...
+
+std::vector<SelectedObject> OpenCSGRenderer::findModelObject(Vector3d near_pt, Vector3d far_pt,int mouse_x, int mouse_y, double tolerance) {
+	std::vector<SelectedObject> result;
+//	printf("opencsg renderer\n");
+//	GLfloat depth=10.0;
+//	GL_CHECKD(glReadPixels(mouse_x, mouse_y, 1, 1, GL_DEPTH_COMPONENT, GL_FLOAT, &depth));
+//	printf("Depth is %g\n",depth);
+
+	return result;
+}
```
Comment (t-paul) [inline src/glview/preview/OpenCSGRenderer.cc:520]: Maybe we can remove this for now and set the measure buttons to disabled?

File: src/gui/MainWindow.cc [hunk 4]
```diff
@@ -2297,12 +2303,138 @@ void MainWindow::actionRenderDone(const shared_ptr<const Geometry>& root_geom)
...
+				double lat;
+				obj1=this->qglview->selected_obj[0];
+				obj2=this->qglview->selected_obj[1];
+				if(obj1.type == SELECTION_POINT && obj2.type == SELECTION_POINT) dist =(obj2.p1-obj1.p1).norm();
+				if(obj1.type == SELECTION_POINT && obj2.type == SELECTION_LINE) dist =calculateLinePointDistance(obj2.p1, obj2.p2,obj1.p1,lat);
+				if(obj1.type == SELECTION_LINE && obj2.type == SELECTION_POINT) dist =calculateLinePointDistance(obj1.p1, obj1.p2,obj2.p1,lat);
+				if(obj1.type == SELECTION_LINE && obj2.type == SELECTION_LINE) dist =calculateLineLineDistance(obj1.p1, obj1.p2,obj2.p1,obj2.p2,lat);
+				if(!isnan(dist)) {
+	    				std::stringstream ss;
+					ss << "Distance is " << fabs(dist);
```
Comment (t-paul) [inline src/gui/MainWindow.cc:2347]: As this is all Qt anyway, I think we should use the Qt string formatting.

File: src/gui/MainWindow.cc [hunk 5]
```diff
@@ -2297,12 +2303,138 @@ void MainWindow::actionRenderDone(const shared_ptr<const Geometry>& root_geom)
...
+{
+	this->qglview->selected_obj.clear();
+	this->qglview->update();
+	this->qglview->measure_state=MEASURE_ANG1;
+}
+
+extern double calculateLineLineDistance(const Vector3d &l1b, const Vector3d &l1e, const Vector3d &l2b, const Vector3d &l2e, double &dist_lat);
+extern double calculateLinePointDistance(const Vector3d &l1, const Vector3d &l2, const Vector3d &pt, double & dist_lat) ;
+
+void MainWindow::leftClick(QPoint mouse)
```
Comment (t-paul) [inline src/gui/MainWindow.cc:2324]: Can we get all the measurement code into a separate class? MainWindow is already way too large.

File: src/gui/QGLView.cc [hunk 6]
```diff
@@ -286,6 +286,11 @@ void QGLView::normalizeAngle(GLdouble& angle)
 void QGLView::mouseMoveEvent(QMouseEvent *event)
 {
   auto this_mouse = event->globalPos();
+  if(measure_state != 0) {
```
Comment (t-paul) [inline src/gui/QGLView.cc:289]: That should use the enum.

**Review by t-paul — CHANGES_REQUESTED** [review]
I'm getting a reproducible crash when starting a measurement before running F6 render. Maybe blocking measure in preview might be the simple answer for now. We can revisit once we have a solution for preview too.

**Comment by gsohler** [comment]
> I'm getting a reproducible crash when starting a measurement before running F6 render. Maybe blocking measure in preview might be the simple answer for now. We can revisit once we have a solution for preview too.

cannot see, but should be blocked by making button inactive

File: src/glview/cgal/CGALRenderer.cc (hunk 2 above)
Comment (gsohler) [reply src/glview/cgal/CGALRenderer.cc:365]: is there any data there ?

**Comment by t-paul** [comment]
Nice, crash issue seems solved, but it gets into some sort of continuous selection mode when not clicking the popup menu but somewhere outside.

The only thing I found so far is using [`hideEvent`](https://doc.qt.io/qt-5/qwidget.html#hideEvent) which unfortunately requires subclassing QMenu and handling of those extra events mentioned in the docs.

File: src/glview/GLView.cc [hunk 7]
```diff
@@ -382,6 +394,49 @@ void GLView::showCrosshairs(const Color4f& col)
   glEnd();
 }
 
+void GLView::showObject(const SelectedObject &obj, const Vector3d &eyedir)
+{
+  auto vd = cam.zoomValue()/200.0;
```
Comment (t-paul) [inline src/glview/GLView.cc:399]: I assume this should keep the markers relative to the model in size regardless of zoom level? For me it starts making the markers quite large when zooming out. Maybe also something to do with orthogonal/perspective? I don't think it's a showstopper, the sizes worked ok with the tests so far.

File: src/glview/cgal/CGALRenderer.cc [hunk 8]
```diff
@@ -330,3 +330,79 @@ BoundingBox CGALRenderer::getBoundingBox() const
...
+	      dist_nearest=dist_lat;
+	      pt1_nearest=ps->vertices[ind1];
+	      pt2_nearest=ps->vertices[ind2];
+	  }
+        }	  
+      }
+   }
+
+  if(!isnan(dist_nearest)) {
+//    printf("Found one line\n");
```
Comment (t-paul) [inline src/glview/cgal/CGALRenderer.cc:399]: Remove or convert to `PRINTDB` maybe with the result values.

File: src/gui/Measurement.cc [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#include <Measurement.h>
```
Comment (t-paul) [inline src/gui/Measurement.cc:27]:
```suggestion
#include "Measurement.h"
```

File: src/gui/QGLView.cc [hunk 10]
```diff
@@ -57,6 +57,7 @@
 #endif
 
 #include "qt-obsolete.h"
+#include <Measurement.h>
```
Comment (t-paul) [inline src/gui/QGLView.cc:60]:
```suggestion
#include "Measurement.h"
```

File: src/gui/Measurement.h [hunk 11]
```diff
@@ -0,0 +1,21 @@
+
+#pragma once
+
+#include <QGLView.h>
```
Comment (t-paul) [inline src/gui/Measurement.h:4]:
```suggestion
#include "QGLView.h"
```

Naming of that file is a bit unfortunate, it's not a Qt header but belongs to OpenSCAD.

**Review by t-paul — CHANGES_REQUESTED** [review]
Added a couple more minor suggestions, looks good in general.

File: src/glview/cgal/CGALRenderer.cc (hunk 2 above)
Comment (gsohler) [reply src/glview/cgal/CGALRenderer.cc]: I am happy to process the vertices coming from CGAL, however,  I was not able to find an usable template code for that. 
we can improve here any time.

**Comment by gsohler** [comment]
> Nice, crash issue seems solved, but it gets into some sort of continuous selection mode when not clicking the popup menu but somewhere outside.
> 
> The only thing I found so far is using [`hideEvent`](https://doc.qt.io/qt-5/qwidget.html#hideEvent) which unfortunately requires subclassing QMenu and handling of those extra events mentioned in the docs.

once distance symbol is selected, you should specify two points. it will stay in measure mode until this happened, maybe we code ESC key to cancel measurement before selecting 2/3 points ?

File: resources/common.qrc.in [hunk 12]
```diff
@@ -130,5 +130,9 @@
     <file>icons/svg-default/vcr-control-step-forward-white.svg</file>
     <file>icons/svg-default/vcr-control-end.svg</file>
     <file>icons/svg-default/vcr-control-end-white.svg</file>
+    <file>icons/svg-default/measure-dist.svg</file>
+    <file>icons/svg-default/measure-dist-white.svg</file>
+    <file>icons/svg-default/measure-ang.svg</file>
+    <file>icons/svg-default/measure-ang-white.svg</file>
```
[the remark this answers is not in the document]
Comment (gsohler) [reply resources/common.qrc.in:136]: yes, there are, but after running the script, *ALL* svg files were modified, so i reverted all files BUT these 4.
Thought that the source is lost, but i was able to restore from the openscad backup directory.

**Comment by t-paul** [comment]
> once distance symbol is selected, you should specify two points. it will stay in measure mode until this happened, maybe we code ESC key to cancel measurement before selecting 2/3 points ?

Oh, I did not consider that case. Yes, that sounds like a good idea.

The case I meant with the comment was completing the measurement such that the menu with the value pops up. Now if you don't click into that menu item but somewhere else, the `finishMeasurement()` is not called but the menu is still closed. So measurement is finished, menu is gone, but it's not reset to idle mode.

File: src/glview/GLView.cc (hunk 7 above)
Comment (gsohler) [reply src/glview/GLView.cc:399]: no subjective issue here. could this relate to QGLView.cc:420

const auto height = dist * tan_degrees(cam.fov ...

**Review by t-paul — APPROVED** [review]
Lets go with that. We can see how to handle the CGAL case along the way...

**Comment by tpimh** [comment]
Just tested it, and it works. There's a small problem though: it only works with rendered models and not previews. If there's nothing to render, measurement is available, but OpenSCAD just crashes when trying to use it. It's obvious why, but something should prevent it from happening (for example, when printing the "no top level geometry to render" warning, also gray out the measuring functions from the menu).

**Comment by drzraf** [comment]
I would to draw a line between two measurement points: If you're measuring the diameter of a circle (and a hole at the center keep you from placing a point to measure the radius), then having a line between both sides of a diameter help visually ensuring it actually crosses the center.

**Comment by andriusr** [comment]
For some reason on MacOS measurement doe not work. I am on 10.14 Mojave. After Render, buttons are enabled, however, when I push them nothing happens. I am not able to select any points on the model.

**Comment by kintel** [comment]
@andriusr I believe this feature only works when the manifold or fast-csg feature is enabled. Could you check if that's the case for you?

**Comment by andriusr** [comment]
Indeed, enabling fast-csg and restarting made it work. Thank you!

**Pull request merged**